### PR TITLE
Performance Proj Part 1: BenchmarkTests

### PR DESCRIPTION
**Pull request by LynneD**
## Overview

Description: Part 1 of performance project is to build benchmark tests to decide whether the new Github branch improve the performance of the system or not. This branch contains benchmark tests for sequencer and CorfuTable. These tests run as drivers to trigger metrics collection and reporting. 

Why should this be merged: 
Start point of Performance Project.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):
1. Start framework for components benchmarks
2. Insert timers at SequencerView, SequencerClient and finish Sequencer benchmarks
3. Write CorfuTable benchmark tests

## TODO
1. add secondary index and cache to CorfuTable benchmark tests
2. record benchmark data for sequencer and CorfuTable and analyze data
3. add benchmark tests for other components.

### CODE REVIEW

File: benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    void runConsumers() {
+        for (int i = 0; i < numThreads; i++) {
+            workers.execute(() -> {
+                try {
+                    long start = System.nanoTime();
+                    Operation op = operationQueue.take();
+                    op.execute();
+                    long latency = System.nanoTime() - start;
+                    System.out.println("nextToken request latency: " + latency);
```
Comment (zhangn49) [inline benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java]: Using System.out.println() is forbidden. Use a logger or output utility. [printLine]

Comment (LynneD) [reply benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java]: Changed it.

File: benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+            if (!finishedInTime) {
+                log.error("not finished in time.");
+                System.exit(1);
+            }
+        } catch (InterruptedException e) {
+            log.error(String.valueOf(e));
+            throw new UnrecoverableCorfuInterruptedError(e);
+        } finally {
+            System.exit(0);
```
Comment (xnull) [inline benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java]: Please don't use System.exit() anywhere exclude `main` method

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package org.corfudb.benchmark;
+
+import org.corfudb.runtime.CorfuRuntime;
+
+/**
+ * BenchmarkTest to CorfuTable.
+ */
+public class CorfuTableBenchmarkTest extends BenchmarkTest {
+    double ratio;
+    String operationName = null;
```
Comment (xnull) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java]: remove null

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    CorfuTableBenchmarkTest(ParseArgs parseArgs) {
+        super(parseArgs);
+        ratio = parseArgs.getRatio();
+        operationName = parseArgs.getOp();
+    }
+
+    private void runProducer() {
+        for (int i = 0; i < numThreads; i++) {
+            CorfuRuntime rt = rts[i % rts.length];
```
Comment (xnull) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java]: rts, why you use an array if arraylist can be sed

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableOperations.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import org.corfudb.util.MetricsUtils;
+
+/**
+ * Operations for CorfuTable, contains build, ,put, get, put+get operations.
+ * TODO: add secondary index and cache, see performance difference, add memory size parameter.
+ */
+@Slf4j
+public class CorfuTableOperations extends Operation {
+    //private Set<String> keySet;
+    private boolean hasSecordaryIndex;
```
Comment (xnull) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableOperations.java]: add final for all fields if you don't update them during the execution

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java  2
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTableOperations.java  5
- benchmark/src/main/java/org/corfudb/benchmark/ParseArgs.java  2
- benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java  4
- benchmark/src/main/java/org/corfudb/benchmark/Streams.java  2
- benchmark/src/main/java/org/corfudb/benchmark/SequencerBenchmarkTest.java  2
- benchmark/src/main/java/org/corfudb/benchmark/Runtimes.java  2
- benchmark/src/main/java/org/corfudb/benchmark/Operation.java  1
- benchmark/src/main/java/org/corfudb/benchmark/SequencerOperations.java  6
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTables.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651458&bid=15577771)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- benchmark/src/main/java/org/corfudb/benchmark/Runtimes.java  2
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTables.java  1
- benchmark/src/main/java/org/corfudb/benchmark/SequencerOperations.java  6
- benchmark/src/main/java/org/corfudb/benchmark/Operation.java  1
- benchmark/src/main/java/org/corfudb/benchmark/BenchmarkTest.java  4
- benchmark/src/main/java/org/corfudb/benchmark/Streams.java  2
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTableOperations.java  5
- benchmark/src/main/java/org/corfudb/benchmark/ParseArgs.java  2
- benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmarkTest.java  2
- benchmark/src/main/java/org/corfudb/benchmark/SequencerBenchmarkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653430&bid=15584321)

**Comment by xnull** [comment]
Obsolete

**Pull request closed without merging**